### ISSUE TITLE
Unioned emty-results wrapped within IDbTransaction throws "System.InvalidOperationException : This MySqlConnection is already in use."

### ISSUE THREAD

**Issue by dalibor-sanzeru** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#764]
### Steps to reproduce
Code splitted into 2 points:
1) Context wrapped in database transaction (dbTransaction.Commit(); is required to simulate exception).
2) All Empty results Unioned repository code - executes Union operation on empty results.

```
// 1)  Context and transaction setup
var options = new DbContextOptionsBuilder<TestDbContext>()          .UseMySql("Server=localhost;Database=testDb;User=root;Password=TestPassword;").Options;
            using (var ctx = new TestDbContext(options))
            {
                using (var dbTransaction = ctx.BeginTransaction())
                {
                    var repo = new ProductRepository(ctx);
                    var res = repo.GetProductSearchHints("This search wont be found in any resultset !!!");
                    ctx.SaveChanges();
                    dbTransaction.Commit();
                }
            }

// 2) Function from repository wont find any result:
        public List<ProductSearchHint> GetProductSearchHints(string search)
        {
            var iProducts = this.dbContext.Products.Where(x => x.Name.StartsWith(search))
                .Select(x => new ProductSearchHint() { ProductSearchType = ProductSearchType.Product, Search = x.Name });

            var iCities = this.dbContext.Cities.Where(x => x.Name.StartsWith(search))
                .Select(x => new ProductSearchHint() { ProductSearchType = ProductSearchType.City, Search = x.Name });

            var iCompanies = this.dbContext.CompanyInfos.Where(x => x.BusinessName.StartsWith(search))
                .Select(x => new ProductSearchHint() { ProductSearchType = ProductSearchType.Company, Search = x.BusinessName });

            var iCategories = this.dbContext.Categories.Where(x => x.Products.Any() && x.Name.StartsWith(search))
                .Select(x => new ProductSearchHint() { ProductSearchType = ProductSearchType.Category, Search = x.Name });

            //THIS CODE IS NOT WORKING !!! Queryable are used (Not ToList() on indivisual unioned queries). All Unioned resultsets have zero results !!! In this case on transaction commit exception is emmited.
            var res = iCategories.Union(iProducts).Union(iCities).Union(iCompanies).ToList();
            return res;

            //THIS UNION CODE BELLOW WITH UNIONs on selected->.ToList() data IS WORKING!!!
            //var resA = iProducts.ToList();
            //var resB = iCities.ToList();
            //var resC = iCompanies.ToList();
            //var resD = iCategories.ToList();

            //var res = resA.Union(resB).Union(resC).Union(resD).ToList();
            //return res;
        }

```

### The issue
Exception is thrown when UNION on all empty results is used within IDbTransaction. When union is rewrite to ToList() individual calls, its working.

If you are seeing an exception, include the full exceptions details (message and stack trace).

```
Exception message:
Message: System.InvalidOperationException : This MySqlConnection is already in use. See https://fl.vu/mysql-conn-reuse

Stack trace:
   at MySqlConnector.Core.ServerSession.StartQuerying(MySqlCommand command) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ServerSession.cs:line 151
   at MySqlConnector.Core.TextCommandExecutor.ExecuteReaderAsync(String commandText, MySqlParameterCollection parameterCollection, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\TextCommandExecutor.cs:line 33
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 276
   at MySql.Data.MySqlClient.MySqlTransaction.CommitAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlTransaction.cs:line 25
   at MySql.Data.MySqlClient.MySqlTransaction.Commit() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlTransaction.cs:line 12
   at Microsoft.EntityFrameworkCore.Storage.RelationalTransaction.Commit()
   at PredavamApi.Controllers.Filters.UnitOfWorkFilterAttribute.OnActionExecuted(ActionExecutedContext context) in C:\Workspace\xxx\yyy\Controllers\Filters\UnitOfWorkFilterAttribute.cs:line 29
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()
```

### Further technical details

MySQL version: MariaDb 10.3.13
Operating system: Windows 10 Proffesional.
Pomelo.EntityFrameworkCore.MySql version: 2.2.0
Microsoft.AspNetCore.App version: Asp.Net core 2.2

**Comment by mguinness** [comment]
It's unfortunate that [Query: Translate IQueryable.Concat/Union/Intersect/Except/etc. to server](https://github.com/aspnet/EntityFrameworkCore/issues/6812) is yet to be implemented meaning that unions are evaluated locally so you may as well use `ToList()` on each query.

**Comment by lauxjpn** [comment]
The https://github.com/aspnet/EntityFrameworkCore/issues/6812 issue has been fixed/implemented for EF Core 3.0.0 and I can confirm that the following code works now as expected:

```c#
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace IssueConsoleTemplate
{
    #region Resultset types

    public enum ProductSearchType
    {
        Product,
        City,
        Company,
        Category
    }

    public class ProductSearchHint
    {
        public ProductSearchType ProductSearchType { set; get; }
        public string Search { set; get; }
    }

    #endregion

    #region Entities

    public class Product
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public int CategoryId { get; set; }
    }

    public class City
    {
        public int Id { get; set; }
        public string Name { get; set; }
    }

    public class CompanyInfo
    {
        public int Id { get; set; }
        public string BusinessName { get; set; }
    }

    public class Category
    {
        public int Id { get; set; }
        public string Name { get; set; }

        public ICollection<Product> Products { get; set; } = new HashSet<Product>();
    }

    #endregion

    public class Context : DbContext
    {
        public DbSet<Product> Products { get; set; }
        public DbSet<City> Cities { get; set; }
        public DbSet<CompanyInfo> CompanyInfos { get; set; }
        public DbSet<Category> Categories { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder
                .UseMySql("server=127.0.0.1;port=3306;user=root;password=;database=Issue764")
                .UseLoggerFactory(LoggerFactory.Create(b => b.AddConsole()))
                .EnableDetailedErrors()
                .EnableSensitiveDataLogging();
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Product>().HasData(new Product
            {
                Id = 1,
                Name = "Chocolate Brownie",
                CategoryId = 1,
            });

            modelBuilder.Entity<City>().HasData(new City
            {
                Id = 1,
                Name = "Chocobo Hills",
            });

            modelBuilder.Entity<CompanyInfo>().HasData(new CompanyInfo
            {
                Id = 1,
                BusinessName = "Bitter Sweet Inc",
            });

            modelBuilder.Entity<Category>(entity =>
            {
                entity.HasMany(e => e.Products)
                    .WithOne()
                    .HasForeignKey(e => e.CategoryId);

                entity.HasData(new Category
                {
                    Id = 1,
                    Name = "Chocolate",
                });
            });
        }
    }

    internal class Program
    {
        private static void Main()
        {
            using var context = new Context();

            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            using var transaction = context.Database.BeginTransaction();

            var results1 = GetProductSearchHints(context, "Vanilla");
            var results2 = GetProductSearchHints(context, "Choco");

            context.SaveChanges();
            transaction.Commit();

            Debug.Assert(results1.Count == 0);
            Debug.Assert(results2.Count == 3);
        }

        private static List<ProductSearchHint> GetProductSearchHints(
            Context dbContext,
            string search)
        {
            var products = dbContext.Products
                .Where(x => x.Name.StartsWith(search))
                .Select(x => new ProductSearchHint
                {
                    ProductSearchType = ProductSearchType.Product,
                    Search = x.Name
                });

            var cities = dbContext.Cities
                .Where(x => x.Name.StartsWith(search))
                .Select(x => new ProductSearchHint
                {
                    ProductSearchType = ProductSearchType.City,
                    Search = x.Name
                });

            var companies = dbContext.CompanyInfos
                .Where(x => x.BusinessName.StartsWith(search))
                .Select(x => new ProductSearchHint
                {
                    ProductSearchType = ProductSearchType.Company,
                    Search = x.BusinessName
                });

            var categories = dbContext.Categories
                .Where(x => x.Products.Any() && x.Name.StartsWith(search))
                .Select(x => new ProductSearchHint
                {
                    ProductSearchType = ProductSearchType.Category,
                    Search = x.Name
                });

            var results = categories
                .Union(products)
                .Union(cities)
                .Union(companies)
                .ToList();

            return results;
        }
    }
}
```

The following fragment shows the generated SQL for the second query:

```sql
Executed DbCommand (2ms) [Parameters=[@__search_0='Choco' (Size = 4000)], CommandType='Text', CommandTimeout='30']

SELECT 3 AS `ProductSearchType`, `c`.`Name` AS `Search`
FROM `Categories` AS `c`
WHERE EXISTS (
    SELECT 1
    FROM `Products` AS `p`
    WHERE `c`.`Id` = `p`.`CategoryId`) AND ((@__search_0 = '') OR (`c`.`Name` IS NOT NULL AND ((`c`.`Name` LIKE CONCAT(@__search_0, '%')) AND ((LEFT(`c`.`Name`, CHAR_LENGTH(CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin)) = CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin) OR (LEFT(`c`.`Name`, CHAR_LENGTH(CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin)) IS NULL AND CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin IS NULL)))))
UNION
SELECT 0 AS `ProductSearchType`, `p0`.`Name` AS `Search`
FROM `Products` AS `p0`
WHERE (@__search_0 = '') OR (`p0`.`Name` IS NOT NULL AND ((`p0`.`Name` LIKE CONCAT(@__search_0, '%')) AND ((LEFT(`p0`.`Name`, CHAR_LENGTH(CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin)) = CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin) OR (LEFT(`p0`.`Name`, CHAR_LENGTH(CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin)) IS NULL AND CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin IS NULL))))
UNION
SELECT 1 AS `ProductSearchType`, `c0`.`Name` AS `Search`
FROM `Cities` AS `c0`
WHERE (@__search_0 = '') OR (`c0`.`Name` IS NOT NULL AND ((`c0`.`Name` LIKE CONCAT(@__search_0, '%')) AND ((LEFT(`c0`.`Name`, CHAR_LENGTH(CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin)) = CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin) OR (LEFT(`c0`.`Name`, CHAR_LENGTH(CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin)) IS NULL AND CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin IS NULL))))
UNION
SELECT 2 AS `ProductSearchType`, `c1`.`BusinessName` AS `Search`
FROM `CompanyInfos` AS `c1`
WHERE (@__search_0 = '') OR (`c1`.`BusinessName` IS NOT NULL AND ((`c1`.`BusinessName` LIKE CONCAT(@__search_0, '%')) AND ((LEFT(`c1`.`BusinessName`, CHAR_LENGTH(CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin)) = CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin) OR (LEFT(`c1`.`BusinessName`, CHAR_LENGTH(CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin)) IS NULL AND CONVERT(@__search_0 USING utf8mb4) COLLATE utf8mb4_bin IS NULL))))
```
